modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/vector/VectorStorage.java

```diff
@@ -32,24 +33,48 @@
     public int size();
 
     /**
+     * Gets element from vector by index and cast it to double.
+     *
      * @param i Vector element index.
      * @return Value obtained for given element index.
      */
     public double get(int i);
 
+    /**
+     * @param i Vector element index.
+     * @param <T> Type of stored element in vector.
+     * @return Value obtained for given element index.
+     */
+    public <T extends Serializable> T getRaw(int i);
+
     /**
      * @param i Vector element index.
      * @param v Value to set at given index.
      */
     public void set(int i, double v);
 
     /**
-     * Gets underlying array if {@link StorageOpsMetrics#isArrayBased()} returns {@code true}.
+     * @param i Vector element index.
+     * @param v Value to set at given index.
+     */
+    public <T extends Serializable> void setRaw(int i, T v);
+
+    /**
+     * Gets underlying array if {@link StorageOpsMetrics#isArrayBased()} returns {@code true} and all values
+     * in vector are Numbers.
      * Returns {@code null} if in other cases.
      *
      * @see StorageOpsMetrics#isArrayBased()
      */
     public default double[] data() {
         return null;
     }
+
+    /**
+     * @return Underlying array of serializable objects {@link StorageOpsMetrics#isArrayBased()} returns {@code true}.
+     * Returns {@code null} if in other cases.
+     */
+    public default Serializable[] rawData() {
+        return null;
```

Comment:
Wouldn't it be better to throw an exception or return empty array?

Reply:
I save such behavior for consistency with old API